// apollo-server-core/dist/utils/dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dispatcher = void 0;
class Dispatcher {
    constructor(targets) {
        this.targets = targets;
    }
    callTargets(targets, methodName, ...args) {
        return targets.map(target => {
            const method = target[methodName];
            if (method && typeof method === 'function') {
                return method.apply(target, args);
            }
        });
    }
    async invokeHookAsync(methodName, ...args) {
        return await Promise.all(this.callTargets(this.targets, methodName, ...args));
    }
    invokeHookSync(methodName, ...args) {
        return this.callTargets(this.targets, methodName, ...args);
    }
    reverseInvokeHookSync(methodName, ...args) {
        return this.callTargets(this.targets.reverse(), methodName, ...args);
    }
    async invokeHooksUntilNonNull(methodName, ...args) {
        for (const target of this.targets) {
            const method = target[methodName];
            if (!(method && typeof method === 'function')) {
                continue;
            }
            const value = await method.apply(target, args);
            if (value !== null) {
                return value;
            }
        }
        return null;
    }
    invokeDidStartHook(methodName, ...args) {
        const didEndHooks = [];
        for (const target of this.targets) {
            const method = target[methodName];
            if (method && typeof method === 'function') {
                const didEndHook = method.apply(target, args);
                if (didEndHook) {
                    didEndHooks.push(didEndHook);
                }
            }
        }
        return (...args) => {
            didEndHooks.reverse();
            for (const didEndHook of didEndHooks) {
                didEndHook(...args);
            }
        };
    }
}
exports.Dispatcher = Dispatcher;
